### ISSUE TITLE
System.InvalidOperationException: Sequence contains no matching element on .NET Framework 4.7 WPF project

### ISSUE THREAD

**Issue by kopfsick** [dotnet/upgrade-assistant#268]
I'm getting Invalid "System.InvalidOperationException: Sequence contains no matching element" when testing the tool on our .NET Framework 4.7 WPF App. It seems to be a problem when determining TFM.

Stacktrace:
<pre><code>Unhandled exception: System.InvalidOperationException: Sequence contains no matching element
   at System.Linq.ThrowHelper.ThrowNoMatchException()
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.DotNet.UpgradeAssistant.MSBuild.MSBuildProject.get_TFM() in /_/src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/MSBuildProject.cs:line 258
   at Microsoft.DotNet.UpgradeAssistant.MSBuild.MSBuildTargetTFMSelector.SelectTFM(IProject project) in /_/src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/MSBuildTargetTFMSelector.cs:line 73
   at Microsoft.DotNet.UpgradeAssistant.Steps.Solution.CurrentProjectSelectionStep.IsCompleted(IUpgradeContext context, IProject project) in /_/src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Solution/CurrentProjectSelectionStep.cs:line 104
   at Microsoft.DotNet.UpgradeAssistant.Steps.Solution.CurrentProjectSelectionStep.<>c__DisplayClass13_0.<IsApplicableImpl>b__0(IProject p) in /_/src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Solution/CurrentProjectSelectionStep.cs:line 42
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.DotNet.UpgradeAssistant.Steps.Solution.CurrentProjectSelectionStep.IsApplicableImpl(IUpgradeContext context) in /_/src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Solution/CurrentProjectSelectionStep.cs:line 42
   at Microsoft.DotNet.UpgradeAssistant.UpgradeStep.IsApplicable(IUpgradeContext context) in /_/src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/UpgradeStep.cs:line 208
   at Microsoft.DotNet.UpgradeAssistant.UpgraderManager.<>c__DisplayClass9_0.<GetStepsForContext>b__0(UpgradeStep s) in /_/src/components/Microsoft.DotNet.UpgradeAssistant/UpgraderManager.cs:line 39
   at System.Linq.Enumerable.WhereListIterator`1.ToList()
   at Microsoft.DotNet.UpgradeAssistant.UpgraderManager.GetNextStepAsync(IUpgradeContext context, CancellationToken token) in /_/src/components/Microsoft.DotNet.UpgradeAssistant/UpgraderManager.cs:line 51
   at Microsoft.DotNet.UpgradeAssistant.Cli.ConsoleUpgrade.RunAsync(CancellationToken token) in /_/src/cli/Microsoft.DotNet.UpgradeAssistant.Cli/ConsoleUpgrade.cs:line 53
   at Microsoft.DotNet.UpgradeAssistant.Cli.ConsoleUpgrade.RunAsync(CancellationToken token) in /_/src/cli/Microsoft.DotNet.UpgradeAssistant.Cli/ConsoleUpgrade.cs:line 96
   at Microsoft.DotNet.UpgradeAssistant.Cli.ConsoleRunner.RunCommandAsync(CancellationToken token) in /_/src/cli/Microsoft.DotNet.UpgradeAssistant.Cli/ConsoleRunner.cs:line 78
   at Microsoft.DotNet.UpgradeAssistant.Cli.ConsoleRunner.StartAsync(CancellationToken token) in /_/src/cli/Microsoft.DotNet.UpgradeAssistant.Cli/ConsoleRunner.cs:line 39
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at System.CommandLine.Invocation.CommandHandler.GetResultCodeAsync(Object value, InvocationContext context)
   at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseErrorReporting>b__21_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseVersionOption>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__22_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseDirective>b__20_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseDebugDirective>b__11_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass14_0.<<UseExceptionHandler>b__0>d.MoveNext()</code></pre>

The project file (a bit redacted):
```xml
<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
  <PropertyGroup>
    <TargetFramework>net47</TargetFramework>
    <RootNamespace>MyCompany.MyApp</RootNamespace>
    <AssemblyName>MyApp</AssemblyName>
    <UseWPF>true</UseWPF>
    <OutputType>WinExe</OutputType>
    <Configurations>Debug;Release;FitNesse;InternalRelease;Beta-RC</Configurations>
    <ApplicationIcon>icon.ico</ApplicationIcon>
    <MyCustomAttribute>true</MyCustomAttribute>
  </PropertyGroup>
  <Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec ContinueOnError="true" Command="&quot;$(SolutionDir)..\Tools\subwcrev.exe&quot; &quot;$(SolutionDir)..\..\..&quot; &quot;$(TargetPath).config&quot; &quot;$(TargetPath).config&quot;" />
    <Exec Command="&quot;$(ProjectDir)AppConfigToIncludeFile.exe&quot; &quot;$(TargetPath).config&quot; &quot;$(TargetPath).config.wxi&quot;" />
  </Target>
  
  <ItemGroup>
    <Content Include="Resources\Images\Splashscreen.png" CopyToOutputDirectory="PreserveNewest" />
    <Content Include="icon.ico" CopyToOutputDirectory="PreserveNewest" />
    <None Include="App.config" TransformOnBuild="true" />
    <None Include="App.*.config" DependentUpon="App.config" IsTransformFile="True" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="Expression.Blend.Sdk" Version="1.0.2" />
    <PackageReference Include="Extended.Wpf.Toolkit" Version="3.6.0" />
    <PackageReference Include="Infragistics.WPF" Version="19.2.90" />
    <PackageReference Include="Infragistics.WPF.Controls.Editors.XamComboEditor" Version="19.2.90" />
    <PackageReference Include="Infragistics.WPF.DataManager" Version="19.2.90" />
    <PackageReference Include="Infragistics.WPF.DataTree" Version="19.2.90" />
    <PackageReference Include="Infragistics.WPF.DockManager" Version="19.2.90" />
    <PackageReference Include="Infragistics.WPF.DragDrop" Version="19.2.90" />
    <PackageReference Include="Infragistics.WPF.Editors" Version="19.2.90" />
    <PackageReference Include="Microsoft.VisualStudio.SlowCheetah" Version="3.2.26">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
    </PackageReference>
  </ItemGroup>

  <ItemGroup Condition="'$(Configuration)' == 'Debug'">
    <!-- ProjectReferences -->
  </ItemGroup>

  <ItemGroup Condition="'$(Configuration)' == 'Debug' or '$(Configuration)' == 'FitNesse'">
    <!-- ProjectReferences -->
  </ItemGroup>

  <ItemGroup Condition="'$(Configuration)' != 'Debug'">
    <!-- ProjectReferences -->
  </ItemGroup>

  <ItemGroup>
    <!-- ProjectReferences -->
  </ItemGroup>
</Project>
```

**Comment by mjrousos** [comment]
Thanks for the report, @kopfsick. At a glance, that csproj looks like something our tool should be able to handle ok. We'll investigate.

**Comment by MatthewSteeples** [comment]
We're also seeing this error on a net48 class library. We have quite a complex csproj file so I started removing chunks until it worked. When I started adding things back in I was able to undo all of my changes (so the file was back to its original state) and the process worked fine...! I don't know if that helps at all (either in diagnosing or a workaround)

**Comment by mjrousos** [comment]
We've seen issues like this in the past when project files were multi-targeted (had the `<TargetFrameworks>` attribute (with an 's' at the end)) since we don't support that yet or in cases where there were heavily customized project files that didn't include a `<TargetFramework>` or `<TargetFrameworkVersion>` attribute at all, making it hard for the tool to know what .NET Framework they were targeting.

@MatthewSteeples, I'm not sure what would have caused the inconsistency you're seeing, but my best guess would be that you have a csproj file with unusual target framework properties.

@kopfsick, I tried reproducing the issue you're running into, but the Upgrade Assistant ran fine for me on the sample csproj you provided. Is it possible that there's another project in your solution without normal target framework properties that might be causing this?

In any case, I've opened #281 to give better error messages when this error happens, so hopefully that can help to at least shed more light on what's going on here. If you have another repro I can try or if the updates in #281 give any more details (once it's merged and published), let me know and I'll take another look.
